### DISCUSSION TITLE
Use cases for this action over official Vercel git integration?

### DISCUSSION THREAD

**Discussion by leerob** [amondnet/vercel-action#126] [General]
Hello! I'd love to hear more about what this action aims to solve that the official Vercel git integration does not.

- What is missing from the official git integration?
- Why did you choose this over the default solution?
- What could we do better?

We'd had some confusion since this is **not a official Vercel solution** and I'd love to have some clarity. Thank you!

https://vercel.com/docs/git

**Comment by vadym-vorobel** [comment]
Hi @leerob! I just want to share my use case for this action with you.

We have our NodeJS application in a private GitHub repository. Also, it is using the NPM package with another private GitHub repository as a source.

To install this package I need an SSH key to be present in the SSH agent. Right now it's not possible with Vercel itself. See more in this issue - https://github.com/vercel/vercel/discussions/5274

So, we've decided to use GitHub Actions + this vercel-action to build the application correctly and publish it to Vercel.

I hope, it makes sense.

**Comment by leerob** [comment]
@vadym-vorobel thank you, that's very helpful 😄

**Comment by dszymczuk** [comment]
Hi @leerob 
My case is to not deploy the application until the tests are passed. If tests pass, I can deploy the application.

**Reply by lewxdev** [reply]
Very new to Vercel, Next.js and setting up CI/CD with GitHub.
Mind enlightening me on how you were able to get this set up? I'm trying to do the exact same thing 😅

**Comment by thedevdavid** [comment]
@leerob I second @dszymczuk 

Also, if I only have 1 member on the Vercel team, none of the PR's made by my GH team members are deployed. I don't want to add my whole team just for the reason of deployment. We're a team of 4. That'd be $80 / month to let them deploy.

**Reply by dgokcin** [reply]
@DLevai94  with that setup, are you able to create multiple custom domains per branch? what I want to achieve is something like ths following ss and I do not know how this can be achieved with this action.

<img width="890" alt="image" src="https://user-images.githubusercontent.com/33603535/202239430-9ac6677d-dad3-4fbc-a305-911298b1847e.png">

If I configure multiple production custom domains, when I deploy with the `--prod` option, even though I use aliases, my qa deployment also points to the production deployment

**Reply by vickkhera** [reply]
This is where I'm stuck, too. Both my preview host and production host are "Production" on this screen. This means any time I deploy to `--prod` both get replaced. I suppose that isn't so bad, as long as you know it is happening. I am deploying prod from a tag off of the main branch, so at that moment both should be the same.

I would really like to know if I can add the domain without having to create a phantom branch to assign to it, or even to link GitHub at all with my Vercel account.

**Comment by euaaaio** [comment]
> * Why did you choose this over the default solution?

Clarity and consistency. When everything related to CI is in one area, has its own structure and is managed by one technology. And not mashed into different interfaces that don't depend on you, work somehow differently, and can change to the worse at any moment.

**Comment by brandonarbini** [comment]
I'm using it to get around the monorepo limit that Vercel has. I have actions that run on changes to each deployable app in Github. Just like that...no more arbitrary limit. Thank you!

**Comment by leo** [comment]
@brandonarbini Thanks for sharing! Would you mind elaborating on which limit you mean exactly?

**Comment by dszymczuk** [comment]
@leo It means that if have monorepo (fe. 3 projects in one folder), you can't initialize 3 independent projects in each folder. You can initialize vercel project only on the root folder.
Check my issue: https://github.com/vercel/vercel/discussions/5294

**Comment by leo** [comment]
@dszymczuk That's an interesting issue – thanks for sharing it! But this issue right here is about why the GitHub Action would be used instead of our [Git Integration](https://vercel.com/docs/git), not about development.

Perhaps @brandonarbini is talking about how every Git repository on Vercel can only be connected to a fixed number of projects within a single Vercel Personal Account or Team?

**Comment by brandonarbini** [comment]
Yep, there are some limitations that I need to get around like the max of 3 projects for personal accounts and 10 for teams. I prefer to work in a monorepo for my personal projects and I have more than 3 there. Also, I just saw the "ignored build step" feature that wasn't there when I was getting things running initially. That will help a lot because I don't like building everything on every push. Ultimately, the limit of 3 on personal is limiting and upgrading to a team plan for $20/mo seems unnecessary since I only deploy changes to these projects very occasionally.

**Comment by leo** [comment]
I see. That makes a lot of sense, thank you! I'll bring this up to the team too.

**Comment by daiyi** [comment]
@leerob Hi there! My setup was too far off the platonic project shape and I was unable to get the build to work with the official git integration to work. 

- **lack of support for private submodules**. The biggest blocker is that I could not find a nice way to get the official integration to authenticate with the dozens of private submodules in my project. I found [this hack](https://github.com/beeinger/vercel-private-submodule) but I didn't like the complexity of adding a shell script. In GitHub Actions I was able to find a simple solution.
- I need to access the preview url for the build step of my project. I couldn't figure out how to do this in Vercel's native GitHub integration. I couldn't figure out how to do this with GitHub actions either, but I was able to find a workaround where I assign an alias based on the PR number after deploying.
- members of my GitHub org can transparently see and modify the build process without needing access to Vercel. I work on a project with rotating contractors coming and going frequently. It's a logistical pain to have to add and remove people to a vercel team who want to work on the build, and remove members in a timely manner as to not incur excessive costs. It's much simpler to move the build into GitHub Actions, and grant permissions one time through GitHub
- I didn't want deploys on every push/commit/branch, that's too much. I couldn't figure out how to turn that off in vercel, but I knew how to limit workflows to PRs only in GitHub Actions.

**Comment by barrymay** [comment]
@leerob Thanks for asking!  After moving my project to a monorepo I needed to disable the git integration I was using for starters.

I second everything @daiyi mentioned, but I'll add a couple things:

- Similar to @daiyi I needed more flexibility with domain management.  The domain manager in the vercel UI is acceptable (if I'm doing the git integration), but it's not sorted well which makes it clumsy, and I really want to organize and check that in somehow.  I know Vercel has been trying to move to a no-config route for ease of use, but for advanced use-cases where automation is key, having good configured setups is really helpful.  Since this is all done by Github Actions, and code-based, it's clear(er) to see what's going on

- For a monorepo, the latest yarn/npm workspace management support by Vercel is not quite there yet.  I was trying upgrading to yarn berry workspaces (failed), then yarn 1 workspaces (worked but the vercel cached build was very slow), and then npm 7 workspaces (which works fine) after I fidgeted with the "Build & Development Settings" output folder.   To get this kind of thing working in Vercel, I need more clear automation and this action definitely helps!

IMO, I think the Vercel CLI is still very helpful (for this kind of automation).  I feel the Vercel "no-config" model works for base cases, but as projects need more features (or integrations with custom API cases), this can get tricky, and a good independent CI/CD deployment model (that can push to Vercel when required) is key. 

To that end,  IMO Vercel could better support advanced cases by allowing developers the flexibility to:
- Set their "Build & Development Settings" via repo config
- Allow deployments of pre-built next packages (for its advanced features, maybe Vercel could allow devs to have an output mode they could hook into when the code gets pushed across?)
- When a user is using the Vercel CI
  - Allow more control over when the git integration runs a build (it really doesn't have to be every build, it just eats unnecessary cycles)
  - Allow easy multi-cancellation of builds on the Vercel UI

**Comment by graup** [comment]
Hi @leerob and @leo

My use case is to deploy the production branch only when there's a tag in git.

**Reply by vickkhera** [reply]
How did you tell Vercel not to overwrite your staging/preview URL? There's no way to remove the "Production" tag on that URL unless it is linked to another branch, which requires me to link the GitHub repo. I was hoping to avoid linking the repo at all and having to make a phantom branch for it.

**Comment by PaulRBerg** [comment]
Suggestion to move this to a GitHub discussion - but @amondnet must activate discussions in this repository first.

GitHub issues should be for specific bugs/ problems.

**Comment by havenchyk** [comment]
@leerob @leo 

I have a monorepo and 2 projects at vercel: client and server. Vercel-git deploys them separately and automatically, but I need to provide server URL to client for HTTP requests. And I didn't find any way to achieve it, except pretending how the server URL will look like, but if the URL is long, it's truncated and some hash is added and I'm not able to replicate this behavior

**Comment by nasso** [comment]
I would use Vercel's integration if it could build my Web Assembly packages (in Rust)

**Comment by pimmee** [comment]
<img width="1287" alt="image" src="https://user-images.githubusercontent.com/33460567/216887220-19b4fec5-5f79-40e3-91f3-5511c02471a1.png">

**Comment by CXwudi** [comment]
My usage is simply, I have a Hugo site and I want the Hugo binary used in development and deployment consist. Right now, I have a `.env` file with `HUGO_VERSION` shared by my both devcontainer and GitHub Action. The GitHub Action builds the Hugo site, and the `vercel-action` simply just deploy the build output.

If I rely on Vercel to handle the build, then when I upgrade Hugo locally, I have to edit the environment variable in Vercel interface manually as well.

**Comment by antony** [comment]
Hi @leerob - just chiming in. Our build process is significantly more complex than just build-this-branch, we deploy to other services that are not vercel.

We monitor all builds for all projects on our github actions infrastructure, so having some builds done via a vercel integration would just cause fragmentation.

I think having a direct git integration is nice for small projects, but when you're a company with a lot of people and employees, it becomes a little limiting.
